### PR TITLE
Feature/notifications

### LINKED ISSUE #398: Add notifications support
Initial version could be just text.
We should explore if icons can be attached across our supported platforms as well.
This needs to hit Windows, macOS, Linux, iOS and Android..!

### PR DESCRIPTION
**Pull request by andydotxyz**
### Description:
Starting work on showing notifications.
This is a work in progress. Currently supports macOS, Linux, Windows and testing.
Adding here for API discussion if required.

To follow: iOS and Android support.

Fixes #398

### Checklist:

- [x] Tests included. <- it is not possible to test the OS delivery of such things, but helper code is tested.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.
- [x] Public APIs match existing style.

### CODE REVIEW

File: app/app_windows.go [hunk 1]
```diff
@@ -29,3 +49,38 @@ func (app *fyneApp) OpenURL(url *url.URL) error {
...
+	if err != nil {
+		fyne.LogError("Could not write script to show notification", err)
+		return
+	}
+	log.Println("PATH", tmpFilePath)
+	defer os.Remove(tmpFilePath)
+
+	cmd := exec.Command("PowerShell", "-ExecutionPolicy", "Bypass", "-File", tmpFilePath)
+	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
+	err = cmd.Run()
```
Comment (stuartmscott) [inline app/app_windows.go:85]:
```suggestion
	err = cmd.Run()
	if err != nil {
		fyne.LogError("Failed to launch windows notify script", err)
	}
```

**Comment by andydotxyz** [comment]
iOS: icon and name from the app just fine :)

08A747C9-F99B-4835-AF6A-7DC485C33AD6_1_105_c

Android: the icon is the fallback app icon - we will need to upgrade how we do packaging to get the icon showing I think.

Screenshot_20200501-164006

File: test/notification.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	"testing"
+
+	"fyne.io/fyne"
+	"github.com/stretchr/testify/assert"
+)
+
+// ExpectNotification allows an app developer to assert that a notification was sent.
+// After the content of f has executed this utility will check that the specified notification was sent.
+func ExpectNotification(t *testing.T, n *fyne.Notification, f func(a fyne.App)) {
+	a := NewApp().(*testApp)
```
[the remark this answers is not in the document]
Comment (toaster) [reply test/notification.go:13]: Why not simply use the current app? When using the `test` package there should be one.

File: app/app_mobile_ios.m [hunk 3]
```diff
@@ -3,9 +3,45 @@
...
+    notifyNum++;
+    NSString *identifier = [NSString stringWithFormat:@"fyne-notify-%d", notifyNum];
+    UNNotificationRequest *request = [UNNotificationRequest requestWithIdentifier:identifier
+        content:content trigger:nil];
+
+    [center addNotificationRequest:request withCompletionHandler:^(NSError * _Nullable error) {
+        if (error != nil) {
+            NSLog(@"Could not send notification: %@", error);
+        }
+    }];
```
Comment (toaster) [inline app/app_mobile_ios.m:30]: I think `content`, `identifier` and `request` need to be released here.
I don't know if exceptions might appear. If so, the release should happen in a `@finally` block.

File: app/app_mobile_ios.m [hunk 4]
```diff
@@ -3,9 +3,45 @@
...
+
+    UNAuthorizationOptions options = UNAuthorizationOptionAlert;
+    [center requestAuthorizationWithOptions:options
+        completionHandler:^(BOOL granted, NSError *_Nullable error) {
+            if (!granted) {
+                NSLog(@"Unable to get permission to send notifications");
+            } else {
+                doSendNotification(center, titleStr, bodyStr);
+            }
+        }];
```
Comment (toaster) [inline app/app_mobile_ios.m:46]: The block should release `titleStr` and `bodyStr`.

File: app/app_mobile_ios.m [hunk 5]
```diff
@@ -3,9 +3,45 @@
...
+        if (error != nil) {
+            NSLog(@"Could not send notification: %@", error);
+        }
+    }];
+}
+
+void sendNotification(char *title, char *body) {
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    NSString *titleStr = [NSString stringWithUTF8String:title];
+    NSString *bodyStr = [NSString stringWithUTF8String:body];
```
Comment (toaster) [inline app/app_mobile_ios.m:36]: suggestion: name the parameters `cTitle` and `cBody` (or `t` and `b`) and remove the `Str` from the NSStrings.

File: app/app_xdg.go [hunk 6]
```diff
@@ -24,6 +26,25 @@ func (app *fyneApp) OpenURL(url *url.URL) error {
...
 }
 
+func (app *fyneApp) SendNotification(n *fyne.Notification) {
+	conn, err := dbus.SessionBus() // shared connection, don't close
+	if err != nil {
+		fyne.LogError("Unable to connect to session D-Bus", err)
+		return
+	}
+
+	appName := ""       // considered fyne.CurrentApp().UniqueID() but really would need a prettier name
```
Comment (toaster) [inline app/app_xdg.go:36]: I think we should not use an empty string here.
Either use the `UniqueID()` (ugly or not) or use a constant like `FyneApp` for now.

File: test/notification.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (andydotxyz) [reply test/notification.go:13]: *test.App is not the app type, it's *test.testApp - i.e. not exported. Not sure I see the benefit in creating a new type just for this callback.
In this instance it is critical that the CurrentApp() is a *test.testApp, so that we can catch the notification.

Inside the test package you could set a different current app for whatever purposes, so creatting a clean one (as we do in some of our tests) seems like the only approach.

File: app/app_darwin.go [hunk 7]
```diff
@@ -33,3 +36,20 @@ func (app *fyneApp) OpenURL(url *url.URL) error {
...
 	return cmd.Run()
 }
+
+func (app *fyneApp) SendNotification(notify *fyne.Notification) {
+	title := escapeNotifyString(notify.Title)
+	content := escapeNotifyString(notify.Content)
+	template := `display notification "%s" with title "%s"`
+	script := fmt.Sprintf(template, content, title)
+
+	err := exec.Command("osascript", "-e", script).Start()
```
[the remark this answers is not in the document]
Comment (andydotxyz) [reply app/app_darwin.go:46]: You can only access UNUserNotificationCenter if you are running a correctly bundled app with appropriate entitlements signed into the application.

I don't think that is a reasonable requirement for fyne app developers and means you can't run them using "go run" or from "go install"ed binaries.

[the remark this answers is not in the document]
Comment (andydotxyz) [reply app/app_darwin.go:46]: I could probably do some checks to verify whether or not this route will succeed - however it means that developers can no longer compile below 10.14... Do we want to do that?

File: app/app_mobile_ios.m (hunk 3 above)
Comment (andydotxyz) [reply app/app_mobile_ios.m:30]: Only content is allocated, fixing.

File: app/app_mobile_ios.m (hunk 4 above)
Comment (andydotxyz) [reply app/app_mobile_ios.m:46]: No, they are created as autoreleased strings - this method does not allocate their memory.

File: test/notification.go [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import (
+	"testing"
+
+	"fyne.io/fyne"
+	"github.com/stretchr/testify/assert"
+)
+
+// ExpectNotification allows an app developer to assert that a notification was sent.
+// After the content of f has executed this utility will check that the specified notification was sent.
+func ExpectNotification(t *testing.T, n *fyne.Notification, f func(a fyne.App)) {
```
[the remark this answers is not in the document]
Comment (andydotxyz) [reply test/notification.go]: Renamed to AssertNotificationSent

File: app/app_darwin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (andydotxyz) [reply app/app_darwin.go:46]: Update on this: Using UNUserNotificationCenter without a proper app bundle crashes hard with a failed assertion that I cannot find how to check for. It's deep in undocumented apple code...

File: test/notification.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartmscott) [reply test/notification.go:17]: Ah, I had forgotten testapp is not exported. Nevertheless I don't think you should create a new testapp here - if any of the dev's test following this call depends on something they've setup in their testapp it will fail in a non-obvious way.

I think it would be better to follow @toaster's suggestion, with an type cast:
```suggestion
	var ta *testApp
	if fyne.CurrentApp() == nil {
		ta = NewApp()
	} else {
		a, ok := fyne.CurrentApp().(*testApp)
		if ok {
			ta = a
		} else {
			t.Fail("Must use test app for notification assertion")
			return
		}
	}
	// Use ta
```

File: app/app_darwin_test.go [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+package app
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestEscapeNotifyString(t *testing.T) {
```
Comment (stuartmscott) [inline app/app_darwin_test.go:13]: nit: test name should match function being tested

File: app/app_mobile_and.c [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+	jmethodID build = find_method(env, cls, "build", "()Landroid/app/Notification;");
+    jobject notif = (*env)->CallObjectMethod(env, builder, build);
+
+	jclass mgrCls = find_class(env, "android/app/NotificationManager");
+    jobject mgr = getSystemService(env, ctx, "notification");
+
+	jmethodID notify = find_method(env, mgrCls, "notify", "(ILandroid/app/Notification;)V");
+    (*env)->CallVoidMethod(env, mgr, notify, nextId, notif);
+    nextId++;
```
Comment (stuartmscott) [inline app/app_mobile_and.c:80]: Lint: Tabs or spaces, not both

File: app/app_windows_test.go [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+package app
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestEscapeNotifyString(t *testing.T) {
```
Comment (stuartmscott) [inline app/app_windows_test.go:13]: nit: ditto

File: test/notification.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (andydotxyz) [reply test/notification.go:17]: I don't really want to fail a test due to internal implementation which could have completed successfully...

[the remark this answers is not in the document]
Comment (stuartmscott) [reply test/notification.go:17]: Fair enough, instead create a new testapp if you can't use the current one, but at least log that this has happened:
```
} else {
    fyne.LogError("Cannot use current app for notification assertion, creating new test app", nil)
    ta = NewApp()
}
```

[the remark this answers is not in the document]
Comment (andydotxyz) [reply test/notification.go:17]: If the concern is the side-effects of setting an app directly then I guess I agree. For that reason - and for the "app.New(); defer app.New()" pattern that is niggling me - I propose something like "UsingApp(fyne.App, func())" utility.

Then this would be:
```
UsingApp(NewApp(), func() {
    ... do stuff to test notification ...
})

func UsingApp(a fyne.App, fn func()) {
    oldApp := fyne.CurrentApp()
    fyne.SetCurrentApp(a)
    fn()
    fyne.SetCurrentApp(oldApp)
}
```

[the remark this answers is not in the document]
Comment (stuartmscott) [reply test/notification.go:17]: Hypothetically, the dev wants to write a test that something triggers a notification only once. They make notification1, call some code that sends it, and finally assert it is sent. Then they make notification2, call some code that should not send it, and finally asserts that the last notification sent to the user was notification1, but instead get "Expected notification1, got nil" <- this is not obvious to the dev at all

[the remark this answers is not in the document]
Comment (andydotxyz) [reply test/notification.go:17]: The method is now named `AssertNotificationSent` - I think it would be expected that this fails due to nil not being the same as notification2.
I'm not sure where the "last notification" comes from - that's not provided by this supporting API.

File: app/app_darwin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (toaster) [reply app/app_darwin.go:46]: Okay, so let it be this way for the first iteration. But you should add an issue for “support OSX notifications from the app instead of Mac script”. You can assign it to me, if you want :).

File: app/app_mobile_ios.m (hunk 3 above)
Comment (toaster) [reply app/app_mobile_ios.m]: So you’re saying that `[NSString stringWithFormat:@"fyne-notify-%d", notifyNum]` doesn’t allocate an `NSString`? Really?

File: app/app_mobile_ios.m (hunk 4 above)
Comment (toaster) [reply app/app_mobile_ios.m]: Where is the autorelease pool?

File: app/app_mobile_ios.m (hunk 3 above)
Comment (toaster) [reply app/app_mobile_ios.m]: Ah okay … autoreleased. But who releases that if there is no pool? Is this the ARC magic?

File: app/app_mobile_ios.m [hunk 12]
```diff
@@ -15,6 +15,7 @@ void openURL(char *urlStr) {
 
 void doSendNotification(UNUserNotificationCenter *center, NSString *title, NSString *body) {
     UNMutableNotificationContent *content = [UNMutableNotificationContent new];
+    [content autorelease];
```
Comment (toaster) [inline app/app_mobile_ios.m:18]: nit: you could chain this :)

File: test/notification.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartmscott) [reply test/notification.go:17]: I just don't think this should create an app, it should either accept App and type cast to testApp, or use fyne.CurrentApp() and throw nil pointer if the dev didn't set one up. That would be less surprising than an Assert creating apps behind the scenes.

[the remark this answers is not in the document]
Comment (andydotxyz) [reply test/notification.go:17]: I doon't understand the concern of creating a new app. This test requires that a testApp be the current app. And to avoid corner cases we also need it to be clean (if the previous notification is the the same as a new expected notification that does not arrive it could pass incorrectly...
